packages/nimble-components/src/table/specs/table-column-sort-hld.md

```diff
@@ -0,0 +1,77 @@
+# Table Column Sorting HLD
+
+## Problem Statement
+
+There is a requirement to have table columns that can be sorted, both programmatically and interactively. This document focuses specifically on programmatic sorting of columns.
+
+### Out of scope of this HLD
+
+While some interactive sorting may be alluded to within this document, the details of interactive sorting are out of scope for this document. For example, this document does not cover:
+
+-   How a column will opt in/out of being interactively sortable
+-   Event(s) related to sorting
+-   Interactions that lead to the sorting state changing, e.g. clicking a header, option(s) in the header's menu
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#874 Programmatically sort columns](https://github.com/ni/nimble/issues/874)
+-   [#854 Interactively sort the columns in a table](https://github.com/ni/nimble/issues/854)
+-   [Table README](./README.md)
+-   [Table Design Doc](https://xd.adobe.com/view/5b476816-dad1-4671-b20a-efe796631c72-0e14/screen/d389dc1e-da4f-4a63-957b-f8b3cc9591b4/specs/)
+
+## Implementation / Design
+
+### API
+
+The sorting state of the table will be declaritively defined within the `TableColumn` elements that are slotted within the table. A column will optionally be able to define its sort direction (i.e. `ascending` vs `descending`) and its sort index. The sort index enables a table to be sorted by multiple columns at the same time. For example, if a table should be sorted by a first name column and then a last name column, the first name column would have a sort index of 0, and the last name column would have a sort index of 1. This is a similar API to AG Grid, which provides [sort](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sort) and [sortIndex](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sortIndex) configurations.
+
+In some cases, a client may want to configure their table to be sorted by a column that is not visible within the table. To enable this, a column within the table can be marked as hidden. A hidden column exists in the table to provide metadata that can be used for operations, such as sorting, but it is not rendered within the table.
+
+A column within the table may not have a one-to-one mapping to a field within the table's data. For example, a hyperlink column might use two fields from the data: a string to use as the hyperlink's display and a string to use as the hyperlink's href. Because of this, a column must define which data field is used when sorting that column. That field is specified by a column definition by implementing an `operationalDataFieldName` getter that returns the name of the field to use for sorting. For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.
```

Comment:
> For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.

I guess the getOperationalDataFieldName could exist in the  getDataRecordFieldNames but it doesn't need to right?

a hyperlink column could have a urlFieldName, displayFieldName, and valueFieldName where only the first two are used to render the cell template and the last one is used for operations like sorting / grouping, etc?

Reply:
That is correct. There is no requirement that the operationalDataFieldName is in the dataRecordFieldNames array.